File: scripts/run_tests.py
#!/usr/bin/env python3
"""
run_tests.py

Description: Python utility script to run tests for the ChefScore Analytics Dashboard
Usage: python scripts/run_tests.py [options]

This script is a Python wrapper around the Node.js test infrastructure,
providing a convenient way to run tests from Python environments.
"""

import argparse
import os
import platform
import subprocess
import sys
from pathlib import Path


# ANSI color codes for prettier output
class Colors:
    RESET = "\033[0m"
    BOLD = "\033[1m"
    RED = "\033[91m"
    GREEN = "\033[92m"
    YELLOW = "\033[93m"
    BLUE = "\033[94m"
    CYAN = "\033[96m"


def log(message, color=""):
    """Print a formatted message to the console."""
    print(f"{color}{message}{Colors.RESET}")


def print_header(text):
    """Print a header with the test type being run."""
    line = "=" * (len(text) + 8)
    print("\n")
    log(line, Colors.CYAN)
    log(f"    {text}    ", Colors.CYAN + Colors.BOLD)
    log(line, Colors.CYAN)
    print("\n")


def run_command(command, cwd=None, shell=True):
    """Run a shell command and return the result."""
    try:
        process = subprocess.run(
            command, cwd=cwd, shell=shell, check=False, text=True, capture_output=True
        )
        return {
            "success": process.returncode == 0,
            "stdout": process.stdout,
            "stderr": process.stderr,
            "returncode": process.returncode,
        }
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "stdout": "",
            "stderr": str(e),
            "returncode": 1,
        }


def check_prerequisites():
    """Check if Node.js and npm are installed."""
    print_header("Checking Prerequisites")

    # Check if Node.js is installed
    node_result = run_command("node --version")
    if not node_result["success"]:
        log("Node.js is not installed or not in the PATH.", Colors.RED)
        log("Please install Node.js from https://nodejs.org/", Colors.YELLOW)
        return False

    node_version = node_result["stdout"].strip()
    log(f"Node.js version: {node_version}", Colors.GREEN)

    # Check if npm is installed
    npm_result = run_command("npm --version")
    if not npm_result["success"]:
        log("npm is not installed or not in the PATH.", Colors.RED)
        log("npm should be included with Node.js installation.", Colors.YELLOW)
        return False

    npm_version = npm_result["stdout"].strip()
    log(f"npm version: {npm_version}", Colors.GREEN)

    return True


def check_dependencies(project_dir):
    """Check if the required dependencies are installed."""
    print_header("Checking Dependencies")

    # Check if package.json exists
    package_json_path = project_dir / "package.json"
    if not package_json_path.exists():
        log(f"package.json not found at {package_json_path}", Colors.RED)
        return False

    # Check if node_modules exists
    node_modules_path = project_dir / "node_modules"
    if not node_modules_path.exists():
        log(
            "node_modules directory not found. Installing dependencies...",
            Colors.YELLOW,
        )
        install_result = run_command("npm install", cwd=project_dir)
        if not install_result["success"]:
            log("Failed to install dependencies:", Colors.RED)
            log(install_result["stderr"], Colors.RED)
            return False
        log("Dependencies installed successfully.", Colors.GREEN)
    else:
        log("node_modules directory found.", Colors.GREEN)

    # Check if Jest is installed
    jest_path = (
        project_dir
        / "node_modules"
        / ".bin"
        / ("jest.cmd" if platform.system() == "Windows" else "jest")
    )
    if not jest_path.exists():
        log("Jest not found in node_modules. Installing Jest...", Colors.YELLOW)
        install_result = run_command(
            "npm install --save-dev jest jest-environment-jsdom @testing-library/jest-dom",
            cwd=project_dir,
        )
        if not install_result["success"]:
            log("Failed to install Jest:", Colors.RED)
            log(install_result["stderr"], Colors.RED)
            return False
        log("Jest installed successfully.", Colors.GREEN)
    else:
        log("Jest found in node_modules.", Colors.GREEN)

    return True


def run_tests(project_dir, test_type=None, coverage=False, watch=False, verbose=False):
    """Run the specified tests."""
    # Determine the command based on options
    if test_type:
        print_header(f"Running {test_type.capitalize()} Tests")
        command = f"npm run test:{test_type}"
    elif coverage:
        print_header("Running Tests with Coverage")
        command = "npm run test:coverage"
    else:
        print_header("Running All Tests")
        command = "npm test"

    # Add watch flag if needed
    if watch:
        command += " -- --watch"

    # Add verbose flag if needed
    if verbose:
        command += " -- --verbose"

    log(f"Executing: {command}", Colors.BOLD)
    print("\n")

    # Run the command
    process = subprocess.Popen(
        command,
        cwd=project_dir,
        shell=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        bufsize=1,
    )

    # Stream the output
    for line in iter(process.stdout.readline, ""):
        sys.stdout.write(line)

    # Wait for the process to complete
    exit_code = process.wait()

    if exit_code == 0:
        log("\nTests completed successfully! 🎉", Colors.GREEN + Colors.BOLD)
        return True
    else:
        log(
            "\nSome tests failed. Please check the output above for details.",
            Colors.RED,
        )
        return False


def parse_arguments():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description="Python utility script to run tests for the ChefScore Analytics Dashboard",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    test_group = parser.add_mutually_exclusive_group()
    test_group.add_argument("--unit", action="store_true", help="Run only unit tests")
    test_group.add_argument(
        "--integration", action="store_true", help="Run only integration tests"
    )
    test_group.add_argument(
        "--e2e", action="store_true", help="Run only end-to-end tests"
    )
    test_group.add_argument(
        "--all", action="store_true", help="Run all tests (default)"
    )
    test_group.add_argument(
        "--coverage", action="store_true", help="Generate coverage report"
    )

    parser.add_argument("--watch", action="store_true", help="Run tests in watch mode")
    parser.add_argument(
        "--verbose", action="store_true", help="Show detailed test output"
    )
    parser.add_argument(
        "--check", action="store_true", help="Only check dependencies, don't run tests"
    )

    return parser.parse_args()


def main():
    """Main function to run the script."""
    # Parse arguments
    args = parse_arguments()

    # Get the project directory (the directory containing this script's parent)
    project_dir = Path(__file__).resolve().parent.parent

    # Print welcome message
    log("🚀 ChefScore Analytics Dashboard Test Runner (Python Edition)", Colors.BOLD)
    log(f"Working directory: {project_dir}", Colors.BLUE)

    # Check prerequisites
    if not check_prerequisites():
        return 1

    # Check dependencies
    if not check_dependencies(project_dir):
        return 1

    # If --check flag is provided, exit after checking dependencies
    if args.check:
        log("Dependency check completed successfully.", Colors.GREEN)
        return 0

    # Determine test type
    test_type = None
    if args.unit:
        test_type = "unit"
    elif args.integration:
        test_type = "integration"
    elif args.e2e:
        test_type = "e2e"

    # Run tests
    success = run_tests(
        project_dir,
        test_type=test_type,
        coverage=args.coverage,
        watch=args.watch,
        verbose=args.verbose,
    )

    return 0 if success else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/run-tests.js
#!/usr/bin/env node

/**
 * run-tests.js
 * 
 * Description: A utility script to run tests for the ChefScore Analytics Dashboard
 * Usage: 
 *   node scripts/run-tests.js [options]
 *   
 * Options:
 *   --unit             Run only unit tests
 *   --integration      Run only integration tests
 *   --e2e              Run only end-to-end tests
 *   --all              Run all tests (default)
 *   --coverage         Generate coverage report
 *   --watch            Run tests in watch mode
 *   --verbose          Show detailed test output
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// ANSI color codes for prettier output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
};

/**
 * Print a formatted message to the console
 */
function log(message, color = '') {
  console.log(`${color}${message}${colors.reset}`);
}

/**
 * Print a header with the test type being run
 */
function printHeader(text) {
  const line = '='.repeat(text.length + 8);
  console.log('\n');
  log(line, colors.cyan);
  log(`    ${text}    `, colors.cyan + colors.bright);
  log(line, colors.cyan);
  console.log('\n');
}

/**
 * Verify that Jest is installed properly
 */
function verifyJestInstallation() {
  try {
    log('Verifying Jest installation...', colors.dim);
    execSync('npx jest --version', { stdio: 'pipe' });
    return true;
  } catch (error) {
    log('Jest not found. Installing dependencies...', colors.yellow);
    try {
      execSync('npm install --save-dev jest @testing-library/jest-dom', { stdio: 'inherit' });
      return true;
    } catch (installError) {
      log('Failed to install Jest. Please run: npm install --save-dev jest @testing-library/jest-dom', colors.red);
      return false;
    }
  }
}

/**
 * Ensure that the Jest configuration exists
 */
function ensureJestConfig() {
  const jestConfigPath = path.join(process.cwd(), 'jest.config.js');
  
  if (!fs.existsSync(jestConfigPath)) {
    log('Creating Jest configuration file...', colors.yellow);
    
    const configContent = `module.exports = {
  testEnvironment: 'jsdom',
  setupFilesAfterEnv: ['<rootDir>/tests/helpers/setup.js'],
  testMatch: [
    '**/tests/**/*.test.js',
    '**/tests/**/*.spec.js'
  ],
  collectCoverageFrom: [
    'js/**/*.js',
    '!**/node_modules/**',
    '!**/vendor/**'
  ],
  coverageDirectory: 'coverage',
  coverageReporters: ['text', 'lcov', 'html'],
  verbose: true
};`;
    
    fs.writeFileSync(jestConfigPath, configContent);
    log('Jest configuration created.', colors.green);
  }
  
  const setupPath = path.join(process.cwd(), 'tests', 'helpers', 'setup.js');
  if (!fs.existsSync(setupPath)) {
    log('Creating Jest setup file...', colors.yellow);
    
    // Create directory if it doesn't exist
    const setupDir = path.dirname(setupPath);
    if (!fs.existsSync(setupDir)) {
      fs.mkdirSync(setupDir, { recursive: true });
    }
    
    const setupContent = `// Jest setup file
const { JSDOM } = require('jsdom');

// Set up a DOM environment for tests that need it
const dom = new JSDOM('<!DOCTYPE html><html><body></body></html>', {
  url: 'http://localhost/',
  pretendToBeVisual: true
});

global.window = dom.window;
global.document = dom.window.document;
global.navigator = dom.window.navigator;
global.localStorage = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};

// Mock for Chart.js
jest.mock('chart.js', () => ({
  Chart: class {
    constructor() {
      return {
        destroy: jest.fn(),
        update: jest.fn()
      };
    }
  }
}));

// Global mocks and setup
global.fetch = jest.fn(() =>
  Promise.resolve({
    ok: true,
    json: () => Promise.resolve({}),
    text: () => Promise.resolve('')
  })
);

// Add custom jest matchers
require('@testing-library/jest-dom');
`;
    
    fs.writeFileSync(setupPath, setupContent);
    log('Jest setup file created.', colors.green);
  }
}

/**
 * Run the specified tests with Jest
 */
function runTests(options) {
  const testTypes = {
    unit: options.unit,
    integration: options.integration,
    e2e: options.e2e,
    all: options.all || (!options.unit && !options.integration && !options.e2e)
  };
  
  const jestArgs = [];
  
  // Configure test pattern based on options
  if (testTypes.all) {
    printHeader('Running All Tests');
  } else {
    const patterns = [];
    if (testTypes.unit) {
      printHeader('Running Unit Tests');
      patterns.push('tests/unit');
    }
    if (testTypes.integration) {
      printHeader('Running Integration Tests');
      patterns.push('tests/integration');
    }
    if (testTypes.e2e) {
      printHeader('Running End-to-End Tests');
      patterns.push('tests/e2e');
    }
    
    if (patterns.length > 0) {
      jestArgs.push(patterns.join('|'));
    }
  }
  
  // Add coverage if requested
  if (options.coverage) {
    jestArgs.push('--coverage');
  }
  
  // Add watch mode if requested
  if (options.watch) {
    jestArgs.push('--watch');
  }
  
  // Add verbose mode if requested
  if (options.verbose) {
    jestArgs.push('--verbose');
  } else {
    // Make the output prettier by default
    jestArgs.push('--colors');
  }
  
  // Build the command
  const command = `npx jest ${jestArgs.join(' ')}`;
  
  try {
    log(`Executing: ${command}`, colors.dim);
    console.log('\n');
    execSync(command, { stdio: 'inherit' });
    
    log('\nTests completed successfully! 🎉', colors.green + colors.bright);
    return true;
  } catch (error) {
    log('\nSome tests failed. Please check the output above for details.', colors.red);
    return false;
  }
}

/**
 * Parse command line arguments
 */
function parseArgs() {
  const args = process.argv.slice(2);
  return {
    unit: args.includes('--unit'),
    integration: args.includes('--integration'),
    e2e: args.includes('--e2e'),
    all: args.includes('--all'),
    coverage: args.includes('--coverage'),
    watch: args.includes('--watch'),
    verbose: args.includes('--verbose')
  };
}

/**
 * Display script usage information
 */
function printUsage() {
  log('\nChefScore Analytics Dashboard Test Runner', colors.bright);
  log('\nUsage:', colors.yellow);
  log('  node scripts/run-tests.js [options]');
  
  log('\nOptions:', colors.yellow);
  log('  --unit             Run only unit tests');
  log('  --integration      Run only integration tests');
  log('  --e2e              Run only end-to-end tests');
  log('  --all              Run all tests (default)');
  log('  --coverage         Generate coverage report');
  log('  --watch            Run tests in watch mode');
  log('  --verbose          Show detailed test output');
  
  log('\nExamples:', colors.yellow);
  log('  node scripts/run-tests.js --unit                  # Run only unit tests');
  log('  node scripts/run-tests.js --coverage              # Run all tests with coverage');
  log('  node scripts/run-tests.js --unit --integration    # Run unit and integration tests');
  console.log('\n');
}

/**
 * Main function to run the script
 */
function main() {
  const options = parseArgs();
  
  // Show usage if --help or -h is provided
  if (process.argv.includes('--help') || process.argv.includes('-h')) {
    printUsage();
    return;
  }
  
  // Verify Jest installation
  if (!verifyJestInstallation()) {
    return;
  }
  
  // Ensure Jest configuration exists
  ensureJestConfig();
  
  // Run the tests
  runTests(options);
}

// Execute the main function
main(); 